syntax/encode.go

```diff
@@ -116,18 +117,45 @@ func marshalerEncoder(e *encodeState, v reflect.Value, opts encOpts) {
 //////////
 
 func uintEncoder(e *encodeState, v reflect.Value, opts encOpts) {
+	if opts.varint {
+		varintEncoder(e, v, opts)
+		return
+	}
+
+	writeUint(e, v.Uint(), int(v.Type().Size()))
+}
+
+func varintEncoder(e *encodeState, v reflect.Value, opts encOpts) {
 	u := v.Uint()
-	switch v.Type().Kind() {
-	case reflect.Uint8:
-		e.WriteByte(byte(u))
-	case reflect.Uint16:
-		e.Write([]byte{byte(u >> 8), byte(u)})
-	case reflect.Uint32:
-		e.Write([]byte{byte(u >> 24), byte(u >> 16), byte(u >> 8), byte(u)})
-	case reflect.Uint64:
-		e.Write([]byte{byte(u >> 56), byte(u >> 48), byte(u >> 40), byte(u >> 32),
-			byte(u >> 24), byte(u >> 16), byte(u >> 8), byte(u)})
+	if (u >> 62) > 0 {
+		panic(fmt.Errorf("uint value is too big for varint"))
+	}
+
+	uintLen := int(v.Type().Size())
+
+	var varintLen int
+	for _, len := range []uint{1, 2, 4, 8} {
+		if u < (uint64(1) << (8*len - 2)) {
+			varintLen = int(len)
+			break
+		}
+	}
+
+	if uintLen < varintLen {
+		panic(fmt.Errorf("Uint too small to fit varint: %d < %d"))
+	}
```

Comment:
How can this happen?

Reply:
It can't.  Vestige of my first draft, which let the caller specify the varint length.